### ISSUE TITLE
OOM on morph-cn node

### ISSUE THREAD

**Issue by fyrchik** [nspcc-dev/neo-go#2063]
I have full logs if needed. There are some RPC errors an hour before and a huge amount of peer connects/disconnects.
While this can be someone trying to exploit fresh VM OOM bugs, I think we should also check network server (this happened on Jun 22).

```
 fatal error: runtime: out of memory

 runtime stack:
 runtime.throw(0x10f1e06, 0x16)
         runtime/panic.go:1117 +0x72
 runtime.sysMap(0xd080000000, 0x4000000, 0x1a65890)
         runtime/mem_linux.go:169 +0xc6
 runtime.(*mheap).sysAlloc(0x1a4c460, 0x400000, 0x7fffffffffff, 0xc00090dee0)
         runtime/malloc.go:729 +0x1e5
 runtime.(*mheap).grow(0x1a4c460, 0x1, 0x0)
         runtime/mheap.go:1346 +0x85
 runtime.(*mheap).allocSpan(0x1a4c460, 0x1, 0xc000902400, 0xccd035b8a0)
         runtime/mheap.go:1173 +0x609
 runtime.(*mheap).alloc.func1()
         runtime/mheap.go:910 +0x59
 runtime.systemstack(0x0)
         runtime/asm_amd64.s:379 +0x66
 runtime.mstart()
         runtime/proc.go:1246

 goroutine 554012 [running]:
 runtime.systemstack_switch()
         runtime/asm_amd64.s:339 fp=0xccd035bb40 sp=0xccd035bb38 pc=0x4702e0
 runtime.(*mheap).alloc(0x1a4c460, 0x1, 0x430124, 0x1a5e210)
         runtime/mheap.go:904 +0x85 fp=0xccd035bb90 sp=0xccd035bb40 pc=0x4280a5
 runtime.(*mcentral).grow(0x1a5e198, 0x0)
         runtime/mcentral.go:232 +0x79 fp=0xccd035bbd8 sp=0xccd035bb90 pc=0x419159
 runtime.(*mcentral).cacheSpan(0x1a5e198, 0x7ff2d144a930)
         runtime/mcentral.go:158 +0x2ff fp=0xccd035bc30 sp=0xccd035bbd8 pc=0x418f3f
 runtime.(*mcache).refill(0x7ff390cf5d28, 0x24)
         runtime/mcache.go:162 +0xaa fp=0xccd035bc78 sp=0xccd035bc30 pc=0x41844a
 runtime.(*mcache).nextFree(0x7ff390cf5d28, 0xd07fffe024, 0x80, 0x80, 0x1)
         runtime/malloc.go:882 +0x8d fp=0xccd035bcb0 sp=0xccd035bc78 pc=0x40d68d
 runtime.mallocgc(0x100, 0x1012640, 0x1, 0xccd035be00)
         runtime/malloc.go:1069 +0x850 fp=0xccd035bd38 sp=0xccd035bcb0 pc=0x40e090
 runtime.newobject(0x1012640, 0xd07fffe080)
         runtime/malloc.go:1177 +0x38 fp=0xccd035bd68 sp=0xccd035bd38 pc=0x40e2f8
 github.com/nspcc-dev/neo-go/pkg/core/block.New(...)
         github.com/nspcc-dev/neo-go/pkg/core/block/block.go:98
 github.com/nspcc-dev/neo-go/pkg/network.(*Message).decodePayload(0xccd035bf10, 0xd07fff6dc0, 0x2b1)
         github.com/nspcc-dev/neo-go/pkg/network/message.go:143 +0x513 fp=0xccd035be10 sp=0xccd035bd68 pc=0xd26c93
 github.com/nspcc-dev/neo-go/pkg/network.(*Message).Decode(0xccd035bf10, 0xccd035bf48, 0xd075dc5b90, 0xccd035bf10)
         github.com/nspcc-dev/neo-go/pkg/network/message.go:119 +0x259 fp=0xccd035be70 sp=0xccd035be10 pc=0xd266b9
 github.com/nspcc-dev/neo-go/pkg/network.(*TCPPeer).handleConn(0xd075dc5b90)
         github.com/nspcc-dev/neo-go/pkg/network/tcp_peer.go:167 +0x259 fp=0xccd035bfd8 sp=0xccd035be70 pc=0xd341b9
 runtime.goexit()
         runtime/asm_amd64.s:1371 +0x1 fp=0xccd035bfe0 sp=0xccd035bfd8 pc=0x471fe1
 created by github.com/nspcc-dev/neo-go/pkg/network.(*TCPTransport).Dial
         github.com/nspcc-dev/neo-go/pkg/network/tcp_transport.go:40 +0x1b1
```

**Comment by fyrchik** [comment]
There about 400 goroutines like this (1293 minutes is certainly not something we would expect):
```
 goroutine 104175 [chan send, 1293 minutes]:
 github.com/nspcc-dev/neo-go/pkg/consensus.(*service).OnPayload(0xc000150140, 0xc000119ea0)
         github.com/nspcc-dev/neo-go/pkg/consensus/consensus.go:391 +0x575
 github.com/nspcc-dev/neo-go/pkg/network.(*Server).handleExtensibleCmd(0xc00249e780, 0xc000119ea0, 0x7, 0xc0061a1c80)
         github.com/nspcc-dev/neo-go/pkg/network/server.go:853 +0x128
 github.com/nspcc-dev/neo-go/pkg/network.(*Server).handleMessage(0xc00249e780, 0x13836c8, 0xc004638fc0, 0xc000959f10, 0x0, 0x0)
```

All of these hang because there is another goroutine which waited for `AddBlock` mutex.
```
 goroutine 277 [semacquire, 1341 minutes]:
 sync.runtime_SemacquireMutex(0xc00009a99c, 0x0, 0x1)
         runtime/sema.go:71 +0x47
 sync.(*Mutex).lockSlow(0xc00009a998)
         sync/mutex.go:138 +0x105
 sync.(*Mutex).Lock(...)
         sync/mutex.go:81
 github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).AddBlock(0xc00009a900, 0xc000326000, 0x0, 0x0)
         github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:536 +0xcd6
 github.com/nspcc-dev/neo-go/pkg/consensus.(*service).processBlock(0xc000150140, 0x137fa00, 0xc000326000)
         github.com/nspcc-dev/neo-go/pkg/consensus/consensus.go:521 +0xbf
 github.com/nspcc-dev/dbft.(*DBFT).checkCommit(0xc0003a9340)
         github.com/nspcc-dev/dbft@v0.0.0-20210302103605-cc75991b7cfb/check.go:73 +0x8b4
 github.com/nspcc-dev/dbft.(*DBFT).onCommit(0xc0003a9340, 0x1381858, 0xc0001705b0)
         github.com/nspcc-dev/dbft@v0.0.0-20210302103605-cc75991b7cfb/dbft.go:494 +0x4c8
 github.com/nspcc-dev/dbft.(*DBFT).OnReceive(0xc0003a9340, 0x1381858, 0xc0001705b0)
         github.com/nspcc-dev/dbft@v0.0.0-20210302103605-cc75991b7cfb/dbft.go:251 +0xef5
 github.com/nspcc-dev/neo-go/pkg/consensus.(*service).eventLoop(0xc000150140)
         github.com/nspcc-dev/neo-go/pkg/consensus/consensus.go:300 +0x7d6
 created by github.com/nspcc-dev/neo-go/pkg/consensus.(*service).Start
         github.com/nspcc-dev/neo-go/pkg/consensus/consensus.go:250 +0xdc
```

**Comment by fyrchik** [comment]
Previous block addition:
```
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | goroutine 100 [chan send, 1341 minutes]:
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | github.com/nspcc-dev/neo-go/pkg/core/mempool.(*Pool).RemoveStale(0xc0002abcc0, 0xc0000e7498, 0x13730d0, 0xc000874640)
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/core/mempool/mem_pool.go:386 +0xae8
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | github.com/nspcc-dev/neo-go/pkg/network.newServerFromConstructors.func1(0x1387320, 0xc00009a900, 0xc0002aa820, 0xc001f9810
0)
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/network/server.go:149 +0x9d
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).storeBlock(0xc00009a900, 0xc001f98100, 0xc0002aa820, 0xc00009a900, 0x0)
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:822 +0x216a
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).AddBlock(0xc00009a900, 0xc001f98100, 0x0, 0x0)
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:579 +0x20f
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | github.com/nspcc-dev/neo-go/pkg/network.(*blockQueue).run(0xc0004d0270)
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/network/blockqueue.go:54 +0x156
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    | created by github.com/nspcc-dev/neo-go/pkg/network.(*Server).Start
Jun 23 06:40:46 souseiseki docker-compose[5251]: neogo-morph-cn    |         github.com/nspcc-dev/neo-go/pkg/network/server.go:273 +0x25d
```

**Comment by roman-khimov** [comment]
Fixed by #2064?

**Comment by fyrchik** [comment]
@roman-khimov should be fixed.
